include/lexy/result.hpp

```diff
@@ -74,18 +74,13 @@ struct _result_storage_non_trivial
     : _has_value(false), _error(LEXY_FWD(args)...)
     {}
 
-#if !defined(__clang__) && defined(__GNUC__) && __GNUC__ == 9
-    // GCC 9 crashes when trying to convert nullopt to a value here.
+    // GCC 9 crashes and MSVC fails to resolve ambiguous overloads when trying to convert nullopt to
+    // a value here.
 
-    template <typename Nullopt,
-              typename = std::enable_if_t<std::is_same_v<std::decay_t<Nullopt>, nullopt>>>
-    _result_storage_non_trivial(result_value_t, Nullopt&&) : _has_value(true), _value()
-    {}
-    template <typename Nullopt,
-              typename = std::enable_if_t<std::is_same_v<std::decay_t<Nullopt>, nullopt>>>
-    _result_storage_non_trivial(result_error_t, Nullopt&&) : _has_value(false), _error()
-    {}
-#endif
+    _result_storage_non_trivial(result_value_t, const nullopt&) : _has_value(true), _value() {}
+    _result_storage_non_trivial(result_value_t, nullopt&&) : _has_value(true), _value() {}
+    _result_storage_non_trivial(result_error_t, const nullopt&) : _has_value(false), _error() {}
+    _result_storage_non_trivial(result_error_t, nullopt&&) : _has_value(false), _error() {}
```

Comment:
Are you sure we won't need `nullopt&` or `const nullopt &&` overloads as well?

Reply:
It would appear it does need those overloads. We could catch all cases if we took `nullopt` by value here, but I assume there's some circular dependency that needs to be avoided. Perhaps it would be better to just revert to what you had before and add msvc to the preprocessor conditions.

Reply:
I'm not sure what you mean by this, could you elaborate?

Reply:
Yeah, I like that option better.